### PR TITLE
linkerd upgrade should require CRDs be installed

### LINKED ISSUE #8373: `linkerd upgrade` should fail if expected CRDs are not present
Start with a stable Linkerd installation:
```
:; curl https://run.linkerd.io/install | sh
:; linkerd version
Client version: stable-2.11.2
Server version: unavailable
:; linkerd install | k apply -f - && linkerd check --wait=60s
```

Then, we upgrade to the edge:
```
:; curl https://run.linkerd.io/install-edge | sh               
:; linkerd version
Client version: edge-22.4.1
Server version: stable-2.11.2
:; linkerd upgrade | k apply -f -
```

`linkerd check` fails, however, because the new CRDs are not installed:

```
:; linkerd version
Client version: edge-22.4.1
Server version: edge-22.4.1
:;  linkerd check --wait=60s                                    
Linkerd core checks
===================

kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ control plane pods are ready
√ cluster networks can be verified
√ cluster networks contains all node podCIDRs

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
× control plane CustomResourceDefinitions exist
    missing CustomResourceDefinitions: authorizationpolicies.policy.linkerd.io, meshtlsauthentications.policy.linkerd.io, networkauthentications.policy.linkerd.io
    see https://linkerd.io/2/checks/#l5d-existence-crd for hints

Status check results are ×
```

I think the appropriate behavior is for `upgrade` to fail when the expected CRDs are not present, instructing the user to run `linkerd upgrade --crds` first. We probably want to validate that the expected CRD _versions_ are present, too; and not just the expected CRD. I.e. we may care that a `v1` is present, but the cluster's version of the CRD may only include `v1beta`.

### PR DESCRIPTION
**Pull request by adleong**
Fixes #8373

We update `CheckCustomResourceDefinitions` so that it not only checks for the existence of the CRDs, but also ensures that they contain the latest version of each CRD.  Note that this means that we'll need to keep this list of CRD versions in `CheckCustomResourceDefinitions` in sync with the actual CRD versions in the templates.  We also add this check to `linkerd upgrade` when the `--crds` flag is not provided.  This means that users who are upgrading will be required to run `linkerd upgrade --crds` first if they don't have the latest version of any of the CRDs.

Signed-off-by: Alex Leong <alex@buoyant.io>

### CODE REVIEW

File: cli/cmd/upgrade.go [hunk 1]
```diff
@@ -156,6 +156,15 @@ func makeUpgradeFlags() *pflag.FlagSet {
 
 func upgradeRunE(ctx context.Context, k *k8s.KubernetesAPI, flags []flag.Flag, crds bool, options valuespkg.Options) error {
 
+	if !crds {
+		err := healthcheck.CheckCustomResourceDefinitions(ctx, k)
+		if err != nil {
+			fmt.Println(err)
+			fmt.Fprintln(os.Stderr, "Linkerd CRDs must be installed first. Run linkerd upgrade with the --crds flag.")
+			os.Exit(1)
```
Comment (olix0r) [inline cli/cmd/upgrade.go]: I don't think `println` makes sense here -- we'll send the error message to output stream, when `upgrade` generates yaml. This means that a command like `linkerd upgrade > linkerd.yaml` could populate the file with non-yaml data if we hit this, for instance.

And I'm pretty sure the caller prints the error message above, so we could do something like
```suggestion
			return fmt.Errorf("Linkerd CRDs must be installed first. Run linkerd upgrade with the --crds flag:\n%w", err)
```

File: pkg/healthcheck/healthcheck_test.go [hunk 2]
```diff
@@ -550,13 +550,13 @@ metadata:
...
-		"serverauthorizations.policy.linkerd.io",
-		"servers.policy.linkerd.io",
-		"serviceprofiles.linkerd.io",
+	for name, version := range map[string]string{
+		"authorizationpolicies.policy.linkerd.io":  "v1alpha1",
+		"meshtlsauthentications.policy.linkerd.io": "v1alpha1",
+		"networkauthentications.policy.linkerd.io": "v1alpha1",
+		"serverauthorizations.policy.linkerd.io":   "v1beta1",
+		"servers.policy.linkerd.io":                "v1beta1",
+		"serviceprofiles.linkerd.io":               "v1alpha2",
```
Comment (olix0r) [inline pkg/healthcheck/healthcheck_test.go:559]: I wonder if there's any way to automate this by the parsing the rendered CRDs?

File: cli/cmd/upgrade.go [hunk 3]
```diff
@@ -169,6 +169,12 @@ func upgradeControlPlaneRunE(ctx context.Context, k *k8s.KubernetesAPI, flags []
 		return err
 	}
 
+	if err = healthcheck.CheckCustomResourceDefinitions(ctx, k); err != nil {
+		fmt.Println(err)
+		fmt.Fprintln(os.Stderr, "Linkerd CRDs must be installed first. Run linkerd upgrade with the --crds flag.")
+		os.Exit(1)
```
Comment (olix0r) [inline cli/cmd/upgrade.go:175]: I don't think `println` makes sense here -- we'll send the error message to output stream, when `upgrade` generates yaml. This means that a command like `linkerd upgrade > linkerd.yaml` could populate the file with non-yaml data if we hit this, for instance.

And I'm pretty sure the caller prints the error message above, so we could do something like
```suggestion
		return fmt.Errorf("Linkerd CRDs must be installed first. Run linkerd upgrade with the --crds flag:\n%w", err)
```

**Review by kleimkuhler — APPROVED** [review]
Looks good! Works well for me

```shell
$ linkerd check
...
× control plane CustomResourceDefinitions exist
    CRD authorizationpolicies.policy.linkerd.io is missing version v1alpha2
    see https://linkerd.io/2/checks/#l5d-existence-crd for hints

Status check results are ×

$ linkerd upgrade
Linkerd CRDs must be installed first. Run linkerd upgrade with the --crds flag:
CRD authorizationpolicies.policy.linkerd.io is missing version v1alpha2
```

**Review by olix0r — APPROVED** [review]

**Pull request merged**